### PR TITLE
lightning: fix generating region job is serialized

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/42456

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/backend/local/local.go [hunk 1]
```diff
@@ -1129,26 +1132,48 @@ func (local *Backend) generateAndSendJob(
...
+	for _, r := range jobRanges {
+		select {
+		case <-egCtx.Done():
+			return eg.Wait()
+		case rangeCh <- r:
 		}
 	}
-	return nil
+	close(rangeCh)
+	return eg.Wait()
```
Comment (gozssky) [inline br/pkg/lightning/backend/local/local.go:1176]: There is a simper concurrency pattern:
```go
eg, egCtx := errgroup.WithContext(ctx)
eg.SetLimit(local.RangeConcurrency)
for _, r := range jobRanges {
    r := r
    eg.Go(func() error {
         ....
     }
}
eg.Wait()
```

**Review by gozssky — APPROVED** [review]

**Comment by lance6716** [comment]
/run-integration-br-test

last test failed on TiDB side

**Review by Benjamin2037 — APPROVED** [review]
LGTM

**Pull request merged**